### PR TITLE
[MXNET-912] Refactoring ctc loss operator

### PR DESCRIPTION
**Pull request by apeforest**
## Description ##
The ctc_loss operator is widely used and has been stable in the past a few releases. This PR is to move this operator from contrib to MXNet regular and refactored the implementation using the new operator interface. 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- Refactoring
- Updated unit test from mx.contrib.ctc_loss to mx.ctc_loss

### CODE REVIEW

**Comment by stu1130** [comment]
Thanks for your contribution @apeforest 
@mxnet-label-bot[pr-work-in-progress]

**Comment by apeforest** [comment]
@szha Refactored the ctc_loss operator using the new interface. Please help to review.

**Comment by apeforest** [comment]
@sandeep-krishnamurthy Please help to change the label to [pr-awaiting-review]. Thanks!

File: src/operator/nn/ctc_loss.cc [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file ctc_loss.cc
+ * \brief CPU Implementation of CTC Loss op
+ */
+#include "./ctc_loss-inl.h"
+#include "../contrib/ctc_include/detail/cpu_ctc.h"
```
Comment (samskalicky) [inline src/operator/nn/ctc_loss.cc]: Can this be moved over too? Seems kindave odd to be referencing something over in contrib.

File: src/operator/nn/ctc_loss.cc [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+namespace mshadow {
```
Comment (samskalicky) [inline src/operator/nn/ctc_loss.cc]: What things int he compute_ctc_cost function are using mshadow? would it be simpler to just put "using namespace mshadow;" in the body of the function?

File: src/operator/nn/ctc_loss.cc [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+template <typename DType>
+ctcStatus_t compute_ctc_cost(const Tensor<cpu, 3, DType> activations,
+                             DType *costs, DType *grads, int *labels,
+                             int *label_lengths, int *data_lengths,
+                             void *workspace, int train, int blank_label) {
```
Comment (samskalicky) [inline src/operator/nn/ctc_loss.cc]: Shouldnt the train variable just be a bool?

File: src/operator/nn/ctc_loss-inl.h [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+    dispatched = storage_type_assign(out_attrs, kDefaultStorage,
+                                     dispatch_mode, DispatchMode::kFCompute);
+  }
+  if (!dispatched) {
+    dispatched = dispatch_fallback(out_attrs, dispatch_mode);
+  }
+  return dispatched;
+}
+
+inline int CTCLossOpNumInputs(const NodeAttrs& attrs) {
```
Comment (samskalicky) [inline src/operator/nn/ctc_loss-inl.h]: can you add a comment to describe how the number of inputs varies? What does it mean when the use_data_lengths or use_label_lengths is set (are they boolean)?

File: src/operator/nn/ctc_loss-inl.h [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+                           std::vector<TShape>* in_attrs,
+                           std::vector<TShape>* out_attrs) {
+    const CTCLossOpParam& param = nnvm::get<CTCLossOpParam>(attrs.parsed);
+    CHECK_EQ(in_attrs->size(), 2U + param.use_data_lengths + param.use_label_lengths);
+    CHECK_EQ(out_attrs->size(), 2U);
+
+    const TShape &dshape = (*in_attrs)[ctc_loss::kData];
+    const TShape &lshape = (*in_attrs)[ctc_loss::kLabel];
+    CHECK_EQ(dshape.ndim(), 3U) << "The data array must be of rank 3.";
+    CHECK_EQ(lshape.ndim(), 2U) << "The labels array must be of rank 2.";
```
Comment (samskalicky) [inline src/operator/nn/ctc_loss-inl.h]: Is "rank" equivalent to "number of dimensions"?

File: src/operator/nn/ctc_loss.cu [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+ * Copyright (c) 2018 by Contributors
+ * \file ctc_loss.cu
+ * \brief GPU Implementation of ctc_loss op
+ */
+
+#include "./ctc_loss-inl.h"
+#include "../contrib/ctc_include/detail/gpu_ctc.h"
+
+namespace mshadow {
+
```
Comment (samskalicky) [inline src/operator/nn/ctc_loss.cu]: same comment as above, would it be simpler to just put "using namespace mshadow;" in the body of the function?

File: src/operator/nn/ctc_loss.cu [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+template <typename DType>
+ctcStatus_t compute_ctc_cost(const Tensor<gpu, 3, DType> activations,
+                             DType *costs, DType *grads, int *labels,
+                             int *label_lengths, int *input_lengths,
+                             void *workspace, int train, int blank_label) {
+  int minibatch = static_cast<int>(activations.size(1));
+  int alphabet_size = static_cast<int>(activations.size(2));
+  mxnet_warpctc::GpuCTC<DType> ctc(alphabet_size, minibatch, workspace,
+                    activations.stream_->stream_, blank_label);
```
Comment (samskalicky) [inline src/operator/nn/ctc_loss.cu]: Where is mxnet_warpctc defined? Any reason to keep it over in contrib directory instead of local here?

**Comment by samskalicky** [comment]
@apeforest Looks good to me, no show stoppers. I added some minor comments

File: src/operator/nn/ctc_loss.cc (hunk 2 above)
Comment (apeforest) [reply src/operator/nn/ctc_loss.cc]: namespace is needed here because compute_ctc_loss function is being called inside ctc_loss-inl.h under namespace mshadow.

File: src/operator/nn/ctc_loss.cu (hunk 6 above)
Comment (apeforest) [reply src/operator/nn/ctc_loss.cu]: namespace is needed here because compute_ctc_loss function is being called inside ctc_loss-inl.h under namespace mshadow.

**Comment by apeforest** [comment]
Thanks to @access2rohit suggestion, added unit test for backward compatibility test. Please review.

File: tests/python/unittest/test_operator.py [hunk 8]
```diff
@@ -4619,6 +4647,85 @@ def check_ctc_loss_grad(blank_label): # from tf
...
+        padding_mask = -1+ (blank_label=='first')
+
+        targets_0 = [0, 1, 2, 1, 0]
+        loss_log_prob_0 = -3.34211
+        input_prob_matrix_0 = np.asarray(
+            [[0.633766, 0.221185, 0.0917319, 0.0129757, 0.0142857, 0.0260553],
+             [0.111121, 0.588392, 0.278779, 0.0055756, 0.00569609, 0.010436],
+             [0.0357786, 0.633813, 0.321418, 0.00249248, 0.00272882, 0.0037688],
+             [0.0663296, 0.643849, 0.280111, 0.00283995, 0.0035545, 0.00331533],
+             [0.458235, 0.396634, 0.123377, 0.00648837, 0.00903441, 0.00623107]],
```
Comment (access2rohit) [inline tests/python/unittest/test_operator.py]: Is it possible to generate them using random no. generators ? or Do we need fixed array values ?

Comment (apeforest) [reply tests/python/unittest/test_operator.py]: This is hand calculated to make sure the ctc_loss operator is implemented correctly. Besides, if using random numbers, the loss values will be different each time. How do we verify the correctness?

Comment (apeforest) [reply tests/python/unittest/test_operator.py]: What do you mean python library? I could not find any standard python library with this loss function except other third-party frameworks.

Comment (access2rohit) [reply tests/python/unittest/test_operator.py]: I see .... previous tests were also manual. My point was if it is possible to write the same functionality in python like a naive approach (it wouldn't be optimal of course, but for sanity check). Since we would want to improve our tests. On the other hand implementation of such logic takes time similar in amount to implementing the operator itself them we can park it in our backlog as todo but having more versatile tests would definitely help :)

**Comment by access2rohit** [comment]
Let's add a TODO for improved tests for CTC loss .... rest looks good to me !

**Comment by apeforest** [comment]
@access2rohit Thanks for your nice suggestions. I have created a ticket https://issues.apache.org/jira/browse/MXNET-1035 to improve function coverage for all operators.
